server/controllers/api/assessmentDayDataController/index.js

```diff
@@ -121,6 +131,56 @@ const AssessmentDayDataController = {
         }
       }
 
+      const isAssessmentInDb = await AssessmentModel.findOne(
+        appDb,
+        assessmentQuery
+      )
+
+      if (!isAssessmentInDb) {
+        const newAssessmentProperties = AssessmentModel.withDefaults({
+          name: assessment,
+        })
+        await AssessmentModel.upsert(
+          appDb,
+          assessmentQuery,
+          newAssessmentProperties
+        )
+      }
+      if (assessment_variables.length) {
+        const assessmentVariablesStream = await AssessmentVariablesModel.all(
+          appDb,
+          assessmentVariableQuery
+        )
+        assessmentVariablesStream.on('data', (variableData) => {
+          const varIndex = assessment_variables.findIndex(
+            (assessmentVar) => assessmentVar.name === variableData.name
+          )
+
+          if (varIndex !== -1) assessment_variables.splice(varIndex, 1)
+        })
+
+        await new Promise((resolve, reject) => {
+          assessmentVariablesStream.on('end', async () => {
+            const areThereNewVars = assessment_variables.length > 0
+
+            if (areThereNewVars) {
+              assessmentVariablesStream.pause()
+
+              await AssessmentVariablesModel.createMany(
+                appDb,
+                assessment_variables
+              )
+
+              assessmentVariablesStream.resume()
+            }
+
+            return resolve()
+          })
+
+          assessmentVariablesStream.on('error', (err) => reject(err))
```

Comment:
I'm trying to understand this logic. Is this making sure we only create new assessment variables and don't create repeats? If so, can we upsert to achieve the same thing?

Reply:
Yes this logic is ensuring that we don't create repeats. The upsert method for mongodb only writes one document at a time and the assessment_variables property is an array of documents, to write an array of documents mongo has an insert many method.

Reply:
would it be a problem to upsert them individually? This seems like a lot of complexity that could be simplified if we did that.

Reply:
[0872fb6](https://github.com/AMP-SCZ/dpdash/pull/726/commits/0872fb6ba250403a2178df163c52f903ce4d1518)